### PR TITLE
Migrate `run-make/no-cdylib-as-rdylib` to `rmake`

### PR DESCRIPTION
**Pull request by Oneirical**
Part of #121876 and the associated [Google Summer of Code project](https://blog.rust-lang.org/2024/05/01/gsoc-2024-selected-projects.html).

> "the test will fail if the cdylib is picked, because it doesn't export any rust symbols"

Is that true? Is there a way to verify?

I suggest maybe extending the test with: (after cleaning the directory)

```rust
    rustc()
        .input("bar.rs")
        .crate_type("cdylib")
        .run();
    rustc()
        .input("foo.rs")
        .prefer_dynamic()
        .run();
    fail();
```

to make sure we're actually testing something here.

### CODE REVIEW

File: src/tools/run-make-support/src/rustc.rs [hunk 1]
```diff
@@ -164,6 +164,12 @@ impl Rustc {
         self
     }
 
+    /// Favours dynamic linking over static linking in rustc. Equivalent to `-C prefer-dynamic``.
+    pub fn prefer_dynamic(&mut self) -> &mut Self {
+        self.cmd.arg("-Cprefer-dynamic");
```
Comment (Nilstrieb) [inline src/tools/run-make-support/src/rustc.rs:169]: I'd prefer not having this helper and just passing `-Cprefer-dynamic` directly

File: tests/run-make/no-cdylib-as-rdylib/rmake.rs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+// This test produces an rlib and a cdylib from bar.rs.
+// Then, foo.rs attempts to link to the bar library.
+// If the test works correctly, rustc will favour the rlib and ignore the cdylib.
```
Comment (jieyouxu) [inline tests/run-make/no-cdylib-as-rdylib/rmake.rs:3]: It's more accurate to say "if the test passes, rustc favored the rlib and ignored the cdylib".

**Comment by jieyouxu** [comment]
Thanks, r=me after squashing the commits into one and after CI is green

@bors delegate+

**Pull request merged**